### PR TITLE
[mlir][Transform] Create a transform interpreter and a preloader pass

### PR DESCRIPTION
**Pull request by nicolasvasilache**
This revision provides the ability to use an arbitrary named sequence op as
the entry point to a transform dialect strategy.

It is also a step towards better transform dialect usage in pass pipelines
that need to preload a transform library rather thanparse it on the fly.

The interpreter itself is significantly simpler than its testing counterpart
by avoiding payload/debug root tags and multiple shared modules.

In the process, the NamedSequenceOp::apply function is adapted to allow it
being an entry point.

NamedSequenceOp is **not** extended to take the PossibleTopLevelTrait at this
time, because the implementation of the trait is specific to allowing one
top-level dangling op with a region such as SequenceOp or AlternativesOp.
In particular, the verifier of PossibleTopLevelTrait does not allow for an
empty body, which is necessary to declare a NamedSequenceOp that gets linked
in separately before application.

In the future, we should dispense with the PossibleTopLevelTrait altogether
and always enter the interpreter with a NamedSequenceOp.

Lastly, relevant TD linking utilities are moved to TransformInterpreterUtils
and reused from there.

### CODE REVIEW

File: mlir/include/mlir/Dialect/Transform/Transforms/Passes.td [hunk 1]
```diff
@@ -42,4 +42,35 @@ def InferEffectsPass : Pass<"transform-infer-effects"> {
...
+    TODO: use a resource blob.
+  }];
+  let options = [
+    ListOption<"transformLibraryPaths", "transform-library-paths", "std::string",
+    "Optional paths to files with modules that should be merged into the "
+    "transform module to provide the definitions of external named sequences.">
+  ];
+}
+
+def InterpreterPass : Pass<"transform-interpreter"> {
```
Comment (matthias-springer) [inline mlir/include/mlir/Dialect/Transform/Transforms/Passes.td:63]: Does this have to be a new pass or could it be an option of the existing `test-transform-dialect-interpreter` pass?

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterUtils.cpp [hunk 2]
```diff
@@ -91,10 +145,51 @@ transform::detail::findTransformEntryPoint(Operation *root, ModuleOp module,
...
+  for (const std::string &libraryFileName : libraryFileNames) {
+    OwningOpRef<ModuleOp> parsedLibrary;
+    auto loc = FileLineColLoc::get(context, libraryFileName, 0, 0);
+    if (failed(transform::detail::parseTransformModuleFromFile(
+            context, libraryFileName, parsedLibrary)))
+      return emitError(loc) << "failed to parse transform library module";
+    parsedLibraries.push_back(std::move(parsedLibrary));
+  }
+
+  // Merge parsed libraries into one module.
```
Comment (matthias-springer) [inline mlir/lib/Dialect/Transform/Transforms/TransformInterpreterUtils.cpp:168]: I think `getPreloadedTransformModule` should be doing that. It currently only takes the first preloaded library and ignores all others. Alternatively, this could also be done `PreloadLibraryPass`: if there is no module preloaded yet, add the first one. On every subsequent run of the `PreloadLibraryPass` pass (and for every file in the list of libraries), merge the symbols.

File: mlir/lib/Dialect/Transform/IR/TransformInterfaces.cpp [hunk 3]
```diff
@@ -60,9 +60,13 @@ ArrayRef<Operation *>
 transform::TransformState::getPayloadOpsView(Value value) const {
   const TransformOpMapping &operationMapping = getMapping(value).direct;
   auto iter = operationMapping.find(value);
-  assert(
-      iter != operationMapping.end() &&
-      "cannot find mapping for payload handle (param/value handle provided?)");
+
+  if (iter == operationMapping.end()) {
+    value.dump();
```
Comment (matthias-springer) [inline mlir/lib/Dialect/Transform/IR/TransformInterfaces.cpp:65]: Leftover debug code

Comment (ftynse) [reply mlir/lib/Dialect/Transform/IR/TransformInterfaces.cpp:65]: Rewrite to `assert(iter != operationMapping.end() && "...")`

File: mlir/lib/Dialect/Transform/IR/TransformOps.cpp [hunk 4]
```diff
@@ -1761,8 +1761,20 @@ DiagnosedSilenceableFailure
...
+  // Note: this is the same implementation as PossibleTopLevelTransformOp but
+  // without attaching the interface / trait since that is tailored to a
+  // dangling top-level op that does not get "called".
+  auto scope = state.make_region_scope(getBody());
+  if (failed(detail::mapPossibleTopLevelTransformOpBlockArguments(
+          state, this->getOperation(), getBody())))
+    return DiagnosedSilenceableFailure::definiteFailure();
+
+  return applySequenceBlock(getBody().front(),
+                            FailurePropagationMode::Propagate, state, results);
```
Comment (ftynse) [inline mlir/lib/Dialect/Transform/IR/TransformOps.cpp:1777]: I don't remember if there is a failure propagation mode on the named sequence (there is one on the include). If there is, let's use it instead of hardcoding here.

File: mlir/include/mlir/Dialect/Transform/Transforms/Passes.td [hunk 5]
```diff
@@ -42,4 +42,35 @@ def InferEffectsPass : Pass<"transform-infer-effects"> {
...
+def PreloadLibraryPass : Pass<"transform-preload-library"> {
+  let summary = "preload transform dialect library";
+  let description = [{
+    This pass preloads a transform library and makes it available to a subsequent
+    transform interpreter passes. The preloading occurs into the Transform
+    dialect and thus provides very limited functionality that does not scale.
+
+    Warning: Only a single such pass should exist for a given MLIR context.
+    This is a temporary solution until a resource-based solution is available.
+    TODO: use a resource blob.
```
Comment (ftynse) [inline mlir/include/mlir/Dialect/Transform/Transforms/Passes.td:54]: I'm not entirely convinced that resources actually solve the problem. Somebody has to own resource...

File: mlir/include/mlir/Dialect/Transform/Transforms/Passes.td (hunk 1 above)
Comment (ftynse) [reply mlir/include/mlir/Dialect/Transform/Transforms/Passes.td:63]: We actually want a non-test pass.

File: mlir/lib/Dialect/Transform/Transforms/InterpreterPass.cpp [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  using Base::Base;
+
+  LogicalResult initialize(MLIRContext *context) override {
+    // TODO: use a resource blob.
+    ModuleOp transformModule =
+        transform::detail::getPreloadedTransformModule(context);
+    if (transformModule) {
+      sharedTransformModule =
+          std::make_shared<OwningOpRef<ModuleOp>>(transformModule.clone());
+    }
```
Comment (ftynse) [inline mlir/lib/Dialect/Transform/Transforms/InterpreterPass.cpp:35]: Do we still need this shared module thing? Is it being mutated?

File: mlir/lib/Dialect/Transform/Transforms/PreloadLibraryPass.cpp [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+} // namespace mlir
+
+namespace {
+class PreloadLibraryPass
+    : public transform::impl::PreloadLibraryPassBase<PreloadLibraryPass> {
+public:
+  using Base::Base;
+
+  LogicalResult initialize(MLIRContext *context) override {
+    OwningOpRef<ModuleOp> mergedParsedLibraries;
```
Comment (ftynse) [inline mlir/lib/Dialect/Transform/Transforms/PreloadLibraryPass.cpp:29]: Can't all this be in `runOnOperation` instead?

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterUtils.cpp [hunk 8]
```diff
@@ -91,10 +145,51 @@ transform::detail::findTransformEntryPoint(Operation *root, ModuleOp module,
...
+  {
+    mergedParsedLibraries.get()->setAttr("transform.with_named_sequence",
+                                         UnitAttr::get(context));
+    IRRewriter rewriter(context);
+    // TODO: extend `mergeSymbolsInto` to support multiple `other` modules.
+    for (OwningOpRef<ModuleOp> &parsedLibrary : parsedLibraries) {
+      if (failed(transform::detail::mergeSymbolsInto(
+              mergedParsedLibraries.get(), std::move(parsedLibrary))))
+        return mergedParsedLibraries->emitError()
+               << "failed to verify merged transform module";
```
Comment (ftynse) [inline mlir/lib/Dialect/Transform/Transforms/TransformInterpreterUtils.cpp:181]: Is it necessarily a verification error? Also, hasn't `mergeSymbolsInto` already emitted a more precise diagnostic?

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterUtils.cpp [hunk 9]
```diff
@@ -91,10 +145,51 @@ transform::detail::findTransformEntryPoint(Operation *root, ModuleOp module,
...
+    return failure();
+
+  // Parse modules from library files.
+  SmallVector<OwningOpRef<ModuleOp>> parsedLibraries;
+  for (const std::string &libraryFileName : libraryFileNames) {
+    OwningOpRef<ModuleOp> parsedLibrary;
+    auto loc = FileLineColLoc::get(context, libraryFileName, 0, 0);
+    if (failed(transform::detail::parseTransformModuleFromFile(
+            context, libraryFileName, parsedLibrary)))
+      return emitError(loc) << "failed to parse transform library module";
```
Comment (ftynse) [inline mlir/lib/Dialect/Transform/Transforms/TransformInterpreterUtils.cpp:164]: Hasn't the parser already emitted an error?

File: mlir/test/Dialect/Transform/Library/lower-to-llvm.mlir [hunk 10]
```diff
@@ -32,8 +32,17 @@ transform.named_sequence @lower_to_cpu(
...
+  //   1. applying reconcile-unrealized-casts on the whole module yields the
+  //      transform applies to transform, when called from a named sequence, at
+  //      this time.
+  //   2. apply_conversion patterns consumes the func but does not produce 
+  //      a new llvm.func.
+  %f6 = transform.structured.match ops{["llvm.func"]} in %module 
+    : (!transform.any_op) -> !transform.any_op
+  %f7 = transform.apply_registered_pass "reconcile-unrealized-casts" to %f6
+    : (!transform.any_op) -> !transform.any_op
+  transform.yield %module : !transform.any_op
```
Comment (ftynse) [inline mlir/test/Dialect/Transform/Library/lower-to-llvm.mlir:45]: Why is this relevant to the patch?

**Review by ftynse — APPROVED** [review]

File: mlir/test/Dialect/Transform/Library/lower-to-llvm.mlir (hunk 10 above)
Comment (nicolasvasilache) [reply mlir/test/Dialect/Transform/Library/lower-to-llvm.mlir:45]: There are 2 things here:
1. some missing tracking that this PR exposes and that I was not able to find yet, I'll file a bug on top to get back to running reconcile-unrealized-casts on module. I am not completely sure how this interacts right now with point 2 below.
2. the other issue is that by dropping `{transform.target_tag="payload"}` support, we end up trying to apply transform to the transform module embedded along with the payload IR. This currently fails and should be fixed separately I believe.

File: mlir/lib/Dialect/Transform/IR/TransformInterfaces.cpp (hunk 3 above)
Comment (nicolasvasilache) [reply mlir/lib/Dialect/Transform/IR/TransformInterfaces.cpp:65]: I wanted to print the actual value before the assert, I think there si a convoluted Twine way but meh.

File: mlir/lib/Dialect/Transform/IR/TransformOps.cpp (hunk 4 above)
Comment (nicolasvasilache) [reply mlir/lib/Dialect/Transform/IR/TransformOps.cpp:1777]: there isn't; I started plumbing it through but overall found it not worth it, we end up only applying this code to the top-level entry point and propagating at this point made sense.

File: mlir/test/Dialect/Transform/Library/lower-to-llvm.mlir (hunk 10 above)
Comment (ftynse) [reply mlir/test/Dialect/Transform/Library/lower-to-llvm.mlir:45]: Okay, I see. How about we keep the copy of the original file running the "test interpreter" pass in parallel to this one.

File: mlir/lib/Dialect/Transform/Transforms/PreloadLibraryPass.cpp (hunk 7 above)
Comment (nicolasvasilache) [reply mlir/lib/Dialect/Transform/Transforms/PreloadLibraryPass.cpp:29]: good point

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterUtils.cpp (hunk 2 above)
Comment (nicolasvasilache) [reply mlir/lib/Dialect/Transform/Transforms/TransformInterpreterUtils.cpp:168]: Good point, in this PR I am just moving some code around, can we keep this for a followup cleanup?

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterUtils.cpp (hunk 8 above)
Comment (nicolasvasilache) [reply mlir/lib/Dialect/Transform/Transforms/TransformInterpreterUtils.cpp:181]: Have not thought about modifying the original impl here, I just moved code.
Let's keep this for a followup PR if we need to restructure things.

File: mlir/test/Dialect/Transform/Library/lower-to-llvm.mlir (hunk 10 above)
Comment (nicolasvasilache) [reply mlir/test/Dialect/Transform/Library/lower-to-llvm.mlir:45]: SG!

File: mlir/include/mlir/Dialect/Transform/Transforms/Passes.td [hunk 11]
```diff
@@ -42,4 +42,35 @@ def InferEffectsPass : Pass<"transform-infer-effects"> {
...
 }
 
+def PreloadLibraryPass : Pass<"transform-preload-library"> {
+  let summary = "preload transform dialect library";
+  let description = [{
+    This pass preloads a transform library and makes it available to a subsequent
+    transform interpreter passes. The preloading occurs into the Transform
+    dialect and thus provides very limited functionality that does not scale.
+
+    Warning: Only a single such pass should exist for a given MLIR context.
```
[the remark this answers is not in the document]
Comment (nicolasvasilache) [reply mlir/include/mlir/Dialect/Transform/Transforms/Passes.td:52]: good point, I'll investigate and see if we can relax that once I plug into a more complex use case like IREE.
For now let's just say I feel safer with a conservative disclaimer.

File: mlir/test/Dialect/Transform/Library/lower-to-llvm.mlir [hunk 12]
```diff
@@ -0,0 +1,48 @@
+// RUN: mlir-opt %s
```
Comment (joker-eph) [inline mlir/test/Dialect/Transform/Library/lower-to-llvm.mlir:1]: This test does not do anything?

**Comment by joker-eph** [comment]
What is the test coverage here? I see passes added and a bunch of code, but no tests to exercise it all?

File: mlir/test/Dialect/Transform/Library/lower-to-llvm.mlir (hunk 12 above)
Comment (nicolasvasilache) [reply mlir/test/Dialect/Transform/Library/lower-to-llvm.mlir:1]: seems I made a mistake and erased the relevant file when renaming indeed, thanks for noticing!
I'll add the missing test file tomorrow and report back here.

Comment (ingomueller-net) [reply mlir/test/Dialect/Transform/Library/lower-to-llvm.mlir:1]: #69190 already adds tests for all I could think of testing but feel free to add more if there's something I missed.

**Comment by nicolasvasilache** [comment]
Added the missing test in ebdb0cbef5e9be57237403c46bfdbe985313bb1c 
What I suspect happened is I forgot to `git push` my latest changes to this branch before merging it manually.
